### ISSUE TITLE
"Package 'text' was not found" when setting a command-line flag

### ISSUE THREAD

**Issue by googleson78** [commercialhaskell/stack#5307]
### General summary/comments (optional)
I'm getting `Package 'text' was not found` when trying to build with `--flag text:integer-simple` on latest lts 14/15.

This is not exclusive to `ghci`.

### Steps to reproduce
```bash

blademaster :: ~ » stack ghci --resolver=lts-13.14 --package text --flag text:integer-simple
<loads fine>
blademaster :: ~ » stack ghci --resolver=lts-15.14 --package text --flag text:integer-simple
Invalid flag specification:
- Package 'text' not found (specified on command line)
```
### Expected
It loads `text` with the `integer-simple` flag set.

### Actual
It doesn't find `text`.

### Stack version
```
blademaster :: ~ » stack --version
Version 2.3.1, Git revision de2a7b694f07de7e6cf17f8c92338c16286b2878 (8103 commits) x86_64 hpack-0.33.0
```

### Method of installation
Via `stack upgrade`.

**Comment by googleson78** [comment]
I would look into this, if I had some clue on where to start, as it is preventing us from moving to `ghc-8.8.3`.

**Comment by qrilka** [comment]
@googleson78 I didn't look deeper into it but I suppose that the main problem here is about the order these options get applied, i.e. flags get checked on the original source map without `--package` taken into account.
I.e. you should check the order https://github.com/commercialhaskell/stack/blob/33c6f6ab16b7526f80d871a3f8e994cd334a438e/src/Stack/Options/PackageParser.hs#L27
and https://github.com/commercialhaskell/stack/blob/33c6f6ab16b7526f80d871a3f8e994cd334a438e/src/Stack/Config.hs#L901
get applied and if it's easy to change it to make `--flag` take `--package` into account

**Comment by googleson78** [comment]
Another thing of note is that
```
blademaster :: ~ » stack ghci --resolver=lts-15.14 --package cryptonite --flag cryptonite:-integer-gmp
```
works fine. (and some other packages too, `text` is the only one I know of with this issue)

@qrilka Thanks! I'll take a look.

**Comment by googleson78** [comment]
Actually I don't think the issue issue stems from `--package` and `--flag` ordering, or at least it's not exclusive to `--package`, because I originally encountered this with `stack build` (it's not a `stack ghci --package` exclusive issue)

**Comment by qrilka** [comment]
I guess you're right and your `cryptonite` note gave me a bit more content - the problem is that `text` is shipped with GHC, e.g. see that it's not in https://github.com/commercialhaskell/stackage-snapshots/blob/master/lts/15/14.yaml (but `cryptonite` is).
Thus you need to override that package which will mean that every package dependent on it will also get rebuilt. If this is possible with the flag `--package` I'm not sure

**Comment by googleson78** [comment]
[After some discussion in `curator`](https://github.com/commercialhaskell/curator/issues/16) it seems to me that the intended solution to "I want to use a flag for `text`" is "add `text` to your `extra-deps` and then specify the flag in your `stack.yaml`".

I think this is not good UX:
* I have to include an `extra-dep` (or make a custom snapshot, in which I still mention the specific version) for something that's bundled with the particular ghc, i.e. with the stackage snapshot (I can even find it on stackage.org if I search for `text`!), meaning I have to be extra careful not to pick the wrong version, and also to remember to check if I need to bump my `text` version whenever I switch snapshots
* It should at least be documented somehow somewhere - it's also a "breaking change". To find out I opened 3 issues in 3 different repos, wasting multiple people's time. Maybe I just haven't found it though.
* It's arguably a functionality "regression", as I could previously use `--flag` to specify `integer-simple` for `text` (although this seems to be more related to design philosophy around snapshots?)

Sorry if I'm coming off as aggressive, I just want some better solution than what I went through while trying to upgrade to a newer lts.
